plugins/launcher_service_plugin/launcher_service_plugin.cpp

```diff
@@ -0,0 +1,1153 @@
+   bool haserror = false, hasok = false;
+   size_t nodecount = _my->_running_clusters[param.cluster_id].nodes.size();
+   std::vector<std::thread> threads;
+   std::vector<std::pair<int, fc::variant> > thread_result;
+   std::vector<std::string> errstrs;
+   thread_result.resize(nodecount);
+   errstrs.resize(nodecount);
+   threads.reserve(nodecount);
+   int k = 0;
+   bool set_errstr = true;
+   for (auto &itr : _my->_running_clusters[param.cluster_id].nodes) {
+      int id = itr.second.id;
+      int port = itr.second.http_port;
+      thread_result[k].first = -1;
+      if (port) {
+         threads.push_back(std::thread([this, k, cid = param.cluster_id, id, port, &thread_result, &hasok, &haserror, &errstrs]() {
+            thread_result[k].first = id;
+            try {
+               thread_result[k].second = _my->get_info(cid, id, true);
+               hasok = true;
+            } catch (boost::system::system_error& e) {
+               haserror = true;
+               errstrs[k] = e.what();
+               std::string url = "http://" + _my->_config.host_name + ":" + _my->itoa(port);
+               thread_result[k].second = fc::mutable_variant_object("error", e.what())("url", url);
+            } catch (fc::exception & e) {
+               haserror = true;
+               errstrs[k] = e.what();
+               std::string url = "http://" + _my->_config.host_name + ":" + _my->itoa(port);
+               thread_result[k].second = fc::mutable_variant_object("error", e.what())("url", url);
+            }  catch (...) {
+               haserror = true;
+               errstrs[k] = "unknown error";
+               std::string url = "http://" + _my->_config.host_name + ":" + _my->itoa(port);
+               thread_result[k].second = fc::mutable_variant_object("error", "unknown error")("url", url);
+            }
+         }));
+      }
+      ++k;
+   }
+   for (auto &t: threads) {
+      t.join();
+   }
+   for (int i = 0; i < k; ++i) {
+      if (thread_result[i].first >= 0) {
+         result[thread_result[i].first] = std::move(thread_result[i].second);
+      }
+   }
+   if (!hasok && haserror) {
+      for (auto &err: errstrs) {
+         if (err.length())
+            throw std::runtime_error(err);
+      }
+   }
+   return fc::mutable_variant_object("result", result);
+}
+
+fc::variant launcher_service_plugin::get_cluster_running_state(launcher_service::cluster_id_param param)
+{
+   if (_my->_running_clusters.find(param.cluster_id) == _my->_running_clusters.end()) {
+      throw std::runtime_error("cluster is not running");
+   }
+   for (auto &itr : _my->_running_clusters[param.cluster_id].nodes) {
+      launcher_service_plugin_impl::node_state &state = itr.second;
+      if (state.child && !state.child->running()) {
+         state.pid = 0;
+         state.child.reset();
+      }
+   }
+   return fc::mutable_variant_object("result", _my->_running_clusters[param.cluster_id]);
+}
+
+#define CATCH_LAUCHER_EXCEPTIONS \
```

Comment:
typo

Reply:
Also why aren't fc exceptions used to be consistent with the rest of the code base?

Reply:
CATCH_LAUCHER_EXCEPTIONS removed.